### ISSUE TITLE
Format support to refer to capturing group contents by name

### ISSUE THREAD

**Issue by trantor** [Genivia/ugrep#119]
Hello.
I've been perusing the documentation for the format directive. I've noticed how one can print out the contents of capturing group using the `%[NUM]#` syntax. However, apart from printing the name of a named capturing group, I see no way to print its contents using the group name as index instead of the numeric index.
Would it be feasible to add this? Is it already there and I've just missed it?

Thanks in advance, especially for such a versatile tool :)

**Comment by genivia-inc** [comment]
Sure. We can extend the `%#` format field for numeric captures `%[NUM]#` to also support capture names with `%[NAME]#`. This means that `(?<foo>PATTERN)` is captured and output with `--format='%[foo]#%~'`. I'll add that to the next update.

**Comment by genivia-inc** [comment]
This can be made even more useful by supporting lists of `NUM` and `NAME` with `%[NUM1|NUM2|...]#` and `%[NAME1|NAME2|...]#` to output the first capturing group pattern that matched.

For example, `%[foo|bar]#` with regex `(?<foo>this)?(?<bar>that)?` will output `this` if `this` matched and if not then output `that` if `that` matched or nothing if neither matched. That may sound complicated, but can be very useful to extract information conditionally from group captures in the order specified in the list.

**Comment by genivia-inc** [comment]
The draft implementation is complete and will be further tested before release. The new and updated FORMAT fields described in the "man page" are:
~~~
FORMAT
       Option  --format=FORMAT  specifies  an  output format for file matches.
       Fields may be used in FORMAT, which expand into the following values:
...
       %[NUM]#
              the regex group capture NUM; requires option -P.

       %[NUM1|NUM2|...]#
              the first group capture NUM that matched; requires option -P.

       %[NAME]#
              the NAMEd group capture; requires option -P and  capturing  pat-
              tern `(?<NAME>PATTERN)', see also %G.

       %[NAME1|NAME2|...]#
              the  first  NAMEd group capture that matched; requires option -P
              and capturing pattern `(?<NAME>PATTERN)', see also %G.

       %G     list of  group  capture  indices/names  that  matched;  requires
              option -P.

       %[TEXT1|TEXT2|...]G
              list  of  TEXT  indexed  by  group capture indices that matched;
              requires option -P.

       %g     the group capture index/name matched or 1; requires option -P.

       %[TEXT1|TEXT2|...]g
              the first TEXT indexed by the first  group  capture  index  that
              matched; requires option -P.
~~~

**Comment by genivia-inc** [comment]
Available in v3.1.12.

**Comment by trantor** [comment]
@genivia-inc Thanks. Can't wait to try it out :D
